### DISCUSSION TITLE
Migration from pytest. is there alternative to FixtureRequest?

### DISCUSSION THREAD

**Discussion by Gelembjuk** [behave/behave#1221] [Q&A] [answered]
Hello.

I started to learn if behave can be used in my project.
I try to adopt some existent pytest tests and create alternatives with Gherkin and behave.

I want to make some steps which will be the replacement for some fixtures from pytest code.
My current problem is that existent fixtures use FixtureRequest object as argument to read options from command line. I would like to keep same functionality, to pass options as command line arguments.
 Existent code is like

```
@pytest.fixture()
def upgrade(pytestconfig: _pytest.fixtures.FixtureRequest)
   # read command line arguments from pytestconfig object
```

Is there any "best practice" to reuse this kind of fixtures? 
In my step i want just call this method, but i need somehow to init object of FixtureRequest class.

```
from behave import *

@given('User upgrades the server')
def step_impl(context):
    """ This calls the upgrade fixture
    """
    request = .... # How to init this object?
    upgrade(request)
```

Any ideas?

**Comment by jenisys** [comment, marked as answer]
`pytest` fixtures and `behave` fixtures are similar in concept.
BUT: You cannot reuse pytest fixtures in `behave`. If you find a way, let me know.

That said:

* Fixture implementations are normally not complex, meaning the heavy lifting is normally outside the fixture function and just used by the fixture function.
* If this is the case, the `behave` fixture implementation should be similar and can reuse this class(es) or function(s), too.

RELATED TO: Usage in a step implementation
* A Fixture is normally needed when you need some setup/teardown/cleanup logic related to a resource.
* Just calling a function will not work in my opinion.
* In `behave`, you should call `use_fixture()` and pass the `context` object to allow the fixture to register a fixture-cleanup functionality.
